### ISSUE TITLE
Compile before require from current directory failing with 8.5

### ISSUE THREAD

**Issue by YaZko** [ProofGeneral/PG#92]
Dear all,

I (finally) switched to Coq 8.5, and am having troubles with PG's compile before require feature. A simple Require Import Test where Test.v is in the current directory fails with the error displayed below. Note that the very same Import does not fail with 8.4 and that I am up to date with the trunk.

As I understand it, -I option has changed quite a bit between 8.4 and 8.5. However, following the documentation, this simple case should not be affected I believe.
* My coq-load-path is set to nil
* No -I options in my coq-prog-name nor coq-prog-args
* My coq-load-path-include-current is set to t (but is anyway obsolete with 8.5 if I understood the documentation correctly)

Am I missing something else, is this a bug or is a _CoqProject now always mandatory to use the CBR feature?

Thanks in advance,
Best,
Yannick

```
-*- mode: compilation; -*-

echo "Require Test." > /var/folders/xq/my2cmz3n26bf9sw4ryqk_cnh0000gn/T/ProofGeneral-coq64463VbT.v /Users/Rainette/.nix-profile/bin/coqdep -Q /Users/Rainette/Work/zakowski/trunk/Thoughts_and_Notes/Simulations/ "" /var/folders/xq/my2cmz3n26bf9sw4ryqk_cnh0000gn/T/ProofGeneral-coq64463VbT.v
*** Warning: in file /var/folders/xq/my2cmz3n26bf9sw4ryqk_cnh0000gn/T/ProofGeneral-coq64463VbT.v, library Test is required and has not been found in the loadpath!
*** Warning: in file /var/folders/xq/my2cmz3n26bf9sw4ryqk_cnh0000gn/T/ProofGeneral-coq64463VbT.v, library Test is required and has not been found in the loadpath!
/var/folders/xq/my2cmz3n26bf9sw4ryqk_cnh0000gn/T/ProofGeneral-coq64463VbT.vo /var/folders/xq/my2cmz3n26bf9sw4ryqk_cnh0000gn/T/ProofGeneral-coq64463VbT.glob /var/folders/xq/my2cmz3n26bf9sw4ryqk_cnh0000gn/T/ProofGeneral-coq64463VbT.v.beautified: /var/folders/xq/my2cmz3n26bf9sw4ryqk_cnh0000gn/T/ProofGeneral-coq64463VbT.v
/var/folders/xq/my2cmz3n26bf9sw4ryqk_cnh0000gn/T/ProofGeneral-coq64463VbT.vio: /var/folders/xq/my2cmz3n26bf9sw4ryqk_cnh0000gn/T/ProofGeneral-coq64463VbT.v
```

**Comment by cpitclaudel** [comment]
Hi Yannick,

Just to clarify the error message that you posted: `/Users/Rainette/Work/zakowski/trunk/Thoughts_and_Notes/Simulations/` is the folder in which both your sources and Test.v are, right?

**Comment by YaZko** [comment]
Hi Clément,

Yes precisely.

**Comment by Matafou** [comment]
Hi,
I can't reproduce this bug.

How do you set the -Q option? Do you have a _coqProject file?

P.

2016-07-15 13:11 GMT+02:00 Yannick Zakowski <notifications@github.com>:

**Comment by YaZko** [comment]
Hi,
I really do not have much more to give you sorry. I really do not do anything fancy, no _CoqProject and nothing particular as far as -Q is concerned (which might be my mistake?).

Starting simply with an empty folder foo at the root, with only two files foo.v and bar.v, the latter requiring the former. Processing 
Require foo.
used to work smoothly with 8.4 (and still does if I switch back without touching anything), but result with 8.5  in a similar error as supra.

In case it could be of any relevance, note that the exact version I use is the 8.5, no pl yet, as distributed in the currently available nix package.

Sorry for the vagueness,
best,
Yannick

-*- mode: compilation; -*-

echo "Require foo." > /var/folders/xq/my2cmz3n26bf9sw4ryqk_cnh0000gn/T/ProofGeneral-coq67678oSV.v /Users/Rainette/.nix-profile/bin/coqdep -Q /Users/Rainette/foo/ "" /var/folders/xq/my2cmz3n26bf9sw4ryqk_cnh0000gn/T/ProofGeneral-coq67678oSV.v
*** Warning: in file /var/folders/xq/my2cmz3n26bf9sw4ryqk_cnh0000gn/T/ProofGeneral-coq67678oSV.v, library foo is required and has not been found in the loadpath!
*** Warning: in file /var/folders/xq/my2cmz3n26bf9sw4ryqk_cnh0000gn/T/ProofGeneral-coq67678oSV.v, library foo is required and has not been found in the loadpath!
/var/folders/xq/my2cmz3n26bf9sw4ryqk_cnh0000gn/T/ProofGeneral-coq67678oSV.vo /var/folders/xq/my2cmz3n26bf9sw4ryqk_cnh0000gn/T/ProofGeneral-coq67678oSV.glob /var/folders/xq/my2cmz3n26bf9sw4ryqk_cnh0000gn/T/ProofGeneral-coq67678oSV.v.beautified: /var/folders/xq/my2cmz3n26bf9sw4ryqk_cnh0000gn/T/ProofGeneral-coq67678oSV.v
/var/folders/xq/my2cmz3n26bf9sw4ryqk_cnh0000gn/T/ProofGeneral-coq67678oSV.vio: /var/folders/xq/my2cmz3n26bf9sw4ryqk_cnh0000gn/T/ProofGeneral-coq67678oSV.v

**Comment by Matafou** [comment]
No probelm we will find out.

There must be a way you tell PG about this -Q option, otherwise pg would
call coqtop with no -I -Q or -R option. Do you know how you add this option?

what happens when you do the following in a terminal:
       /Users/Rainette/.nix-profile/bin/coqdep -Q /Users/Rainette/foo/ ""
<yourfile>.v
where <yourfile>.v is the file containing the Require.

Do you get the same error? (if yes it may be a coq v8.5 issue).

Please check also you start with no old .vo in your directory (compiled
with earlier verisons of coq).

P.


2016-07-20 13:27 GMT+02:00 Yannick Zakowski <notifications@github.com>:

**Comment by YaZko** [comment]
I must say I do not know where this -Q might come from. As far as I know, the only things I configure are in my .spacemacs, i.e.:
* loading PG and company-coq
* a few graphical things
  (setq proof-three-window-enable t)
  (setq proof-three-window-mode-policy 'hybrid)
  (setq proof-splash-enable nil)
* enabling coq-compile-before-require
* setting a hook for personal bindings

Calling coqdep (without any .vo left) from a terminal seems to work smoothly:

> /Users/Rainette/.nix-profile/bin/coqdep -Q /Users/Rainette/foo/ "" bar.v 
> bar.vo bar.glob bar.v.beautified: bar.v /Users/Rainette/foo/foo.vo
> bar.vio: bar.v /Users/Rainette/foo/foo.vio

Yannick

**Comment by Matafou** [comment]
Very strange.
Can you execute the same coqdep command from emacs (M-! + paste the same command). Does it work too?

**Comment by Matafou** [comment]
Also try to set coq-load-path-include-current to nil.

**Comment by YaZko** [comment]
Executing coqdep from emacs worked smoothly the same way while still having the error. 
However setting coq-load-path-include-current to nil did solve the problem, thanks!
I am a bit confused since I thought to have given it a shot earlier, but I must have messed it up I guess.

Should we wish for this option to be enable by default?

Thanks again for your time.
Y

**Comment by Matafou** [comment]
We should still understand why the -Q option for coqdep failed. It
shouldn't.

When you have time, please try again with v8.5pl1 (or more recent).


2016-07-20 17:51 GMT+02:00 Yannick Zakowski <notifications@github.com>:

**Comment by YaZko** [comment]
Hi Pierre,

I am terribly sorry for the delay, I went on holidays and completely forgot to get back to this. 
So I just gave a shot with the 8.5pl2, and a new fresh install of PG, I still get the exact same error if I do not specify  (setq coq-load-path-include-current nil). If I do, once again, it works smoothly.

I hardly see how it could be relevant, but please note that I use Spacemacs.

Best,
Y

**Comment by Matafou** [comment]
By looking at the coq-system.el file I realize that there is some code to detect coq version (coq-autodetect-version) but it is not really plugged in functions like coq-prog-args.

@cpitclaudel can you have a look at this part please? I can't see why option -Q would appear in a empty environment but I may miss something.

P.

**Comment by YaZko** [comment]
Hey,
I tried to take a look out of curiosity. There is a bit too much things going on for me to conclude, but here is what I gather if it may help:

- coq-seq-preprocess-require-commands calls coq-seq-check-module calls coq-seq-map-module-id-to-obj-file
- This guy figures that coq-load-path-include-current is true, hence append to the previously empty coq-load-path the current directory, and feed it to coq-seq-get-library-dependencies
- The latter will dramatically finds a -Q option when executing (coq-include-options coq-load-path (file-name-directory lib-src-file) (coq--pre-v85))) where coq-load-path is still the list reduced to the current directory and coq--pre-v85 is nil.
- coq-include-options iterates on the list, but therefore simply reduces to a call to coq-option-of-load-path-entry with the same arguments
- And finally here the current directory is patterned matched by (and (pred stringp) dir), therefore appending the -Q which leads right after in coq-seq-get-library-dependencies to the failure of coqdep

So as far as I can tell, the bug seems to be that detecting a post 8.5 version should set coq-load-path-include-current to false. Otherwise in coq-option-of-load-path-entry, coq-load-path-include-current being set to true by default, it will always be called on the current directory and systematically add a -Q option to it.

However I am discovering the code of PG right now so my apologies if I only feed you with trivialities or erroneous remarks.
Best,
Y

**Comment by Blaisorblade** [comment]
Having the same problem (and same fix) *after I switched from ProofGeneral 4.2 to the current version*.

BTW, `M-x customize-group coq` shows uninitialized values on current master (25355560b4aa25b17b754c33c9ba2a18d7498df4):
```
Creating customization items...
Creating group...
Creating group entries...
custom-variable-value-create: Invalid function: "\"Private_\" \"_subproof\""
```

**Comment by hendriktews** [comment]
I had the same problem when I tried 8.5 this week. PR #118 should fix this.
